src/api/store/ws.sagas.test.ts

```diff
@@ -0,0 +1,218 @@
+          gasLimit: '0x1',
+          gasPrice: '0x1',
+          data: '0x',
+          value: '0x1',
+          chainId: 3
+        }
+      ]
+    };
+
+    expect(validateRequest(JSON.stringify(request))).toStrictEqual([null, request]);
+  });
+});
+
+describe('waitForResponse', () => {
+  it('returns the payload if the request ID matches', async () => {
+    await expectSaga(waitForResponse, 1)
+      .take(reply)
+      .dispatch(reply({ id: 1, result: 'foo' }))
+      .returns({ id: 1, result: 'foo' })
+      .silentRun();
+
+    await expectSaga(waitForResponse, 1)
+      .take(reply)
+      .dispatch(reply({ id: 2, result: 'foo' }))
+      .not.returns({ id: 1, result: 'foo' })
+      .silentRun();
+  });
+});
+
+describe('handleRequest', () => {
+  const socket = ({
+    send: jest.fn()
+  } as unknown) as WebSocket;
+
+  const request = ({
+    headers: {
+      origin: 'https://app.mycrypto.com/foo'
+    }
+  } as unknown) as IncomingMessage;
+
+  it('sends an error on invalid request', async () => {
+    await expectSaga(handleRequest, { socket, request, data: '' }).silentRun();
+
+    expect(socket.send).toHaveBeenCalledWith(
+      JSON.stringify({
+        jsonrpc: '2.0',
+        id: null,
+        error: { code: '-32700', message: 'Parse error' }
+      })
+    );
+  });
+
+  it('puts a method request and sends the result', async () => {
+    const { params: _, ...accountsRequest } = createJsonRpcRequest(JsonRPCMethod.Accounts);
+    await expectSaga(handleRequest, { socket, request, data: JSON.stringify(accountsRequest) })
+      .put(requestAccounts({ origin: fRequestOrigin, request: accountsRequest }))
+      .call(waitForResponse, accountsRequest.id)
+      .dispatch(
+        reply({
+          id: accountsRequest.id,
+          result: [fAccount.address]
+        })
+      )
+      .silentRun();
+
+    expect(socket.send).toHaveBeenCalledWith(
+      JSON.stringify({
+        jsonrpc: '2.0',
+        id: accountsRequest.id,
+        result: [fAccount.address]
+      })
+    );
+
+    await expectSaga(handleRequest, { socket, request, data: JSON.stringify(fTxRequest) })
+      .put(requestSignTransaction({ origin: fRequestOrigin, request: fTxRequest }))
+      .call(waitForResponse, fTxRequest.id)
+      .dispatch(
+        reply({
+          id: fTxRequest.id,
+          result: fSignedTx
+        })
+      )
+      .silentRun();
+
+    expect(socket.send).toHaveBeenCalledWith(
+      JSON.stringify({
+        jsonrpc: '2.0',
+        id: fTxRequest.id,
+        result: fSignedTx
+      })
+    );
+  });
+});
+
+describe('requestWatcherWorker', () => {
+  it('handles socket requests', async () => {
+    const server = WebSocket.Server as jest.MockedClass<typeof WebSocket.Server>;
+    const saga = expectSaga(requestWatcherWorker);
+    const promise = saga.silentRun();
+
+    const socket = {
+      send: jest.fn(),
+      on: jest.fn()
+    };
+
+    const request = {
+      origin: fRequestOrigin
```

Comment:
If this is the header, then `fRequestOrigin` is not a valid one. Since it doesn't include `https://` etc.